### DISCUSSION TITLE
TUYA: ME81H and PID control

### DISCUSSION THREAD

**Discussion by Kolia56** [arendst/Tasmota#19836] [Q&A]
Hello,
Is there any way to get PID control module working on a Tuya MCU based thermostat as ME81H is (and many of them). If no, what would be the units that are ESP32 based with same kind of look and fell as ME81H? I have two of these units on which I replaced the original MCU by an ESP32 flashed with Tasmota 12.4.0 and that are working well with HA. However, having PID control on it would be a really great.
Thank you

**Comment by barbudor** [comment]
If I understand well, you already have those units running with Tasmota32
As far as you have a temperature sensor and an power control you should be able to use PID

Have you already tested Tasmota thermostat feature?
If yes, why are you not happy with it and would prefer PID?

**Comment by Kolia56** [comment]
Thanks @barbudor . Indeed I have a working unit connected to HA. Just to let you know, 4 software thermostat are currently running in HA using [HASmartThermostat] using PID control (https://github.com/ScratMan/HASmartThermostat), but I'd like to let physical units to do the regulation job, seems to me it is more efficient.

One very first question. Current unit is loaded with a compiled version where #define USE_TIMEPROP and #define USE_PID  are active. Shall I instead deactivate them and activate #define USE_THERMOSTAT ?

Temperature sensor is the one embedded in the unit, so I guess I have to issue.
```SensorInputSet 1```
in order to get the Hybrid controller working.

But there are some of the commands that I don't understand ([doc](https://tasmota.github.io/docs/Commands/#thermostat)). For instance
Enables or disables the physical output for instance
```
EnableOutputSet
``` 
does not seem to produce any effect. If it set to 0, I can here the relay switching when temperature is above/below target.
Another one, Sets the minimum time in minutes within a cycle for the PI controller to switch off the output, below it, it will stay on. Current value is set to 0 that is: 3 minutes
```
TimeMinTurnoffActionSet
```
When the target temp is set to a value above/below current I can here the relay switch well before the 3 minutes delay.

The physical unit provides an external temperature sensor input. 

There must be something quite obvious I don't understand, hopefully you will bring me light.
One last word: what is the best approach to define parameters of the hybrid controller. These units have a temperature measurement resolution of 0.5°C. 

Thank you for your help

**Comment by Kolia56** [comment]
Since then I made some progress. I recompiled with the thermostat feature and I get a nice thermostat running.  I read extensively documentation, however, looking at the interface, I have still some questions:

- To what is Current Temperature related to?
- What is the gradient temperature? I see there is a command to read the value, but I could not see any reference to it in the thermostat documentation.
- is there a way to correct display of Temperature set?

![image](https://github.com/arendst/Tasmota/assets/20872098/8e862399-7199-41ef-8bf0-433c412350de)

Thanks again

**Reply by wlan4you** [reply]
Hello,
please can you send the Template for that?
Thank you

**Reply by nikvostro** [reply]
> Hello, please can you send the Template for that? Thank you

https://templates.blakadder.com/ME81H.html

**Reply by wlan4you** [reply]
No, that displayed not all this information like the Screenshot.
I have tryed your link before

**Reply by wlan4you** [reply]
Ohhh or i have the wrong Firmware?
I choosed the original: http://ota.tasmota.com/tasmota/release/tasmota.bin.gz in V14.3.0

**Reply by sfromis** [reply]
As Kolia56 mentioned, that was after a custom build with the thermostat feature.

**Comment by Kolia56** [comment]
After a lot of expriment and reading, I understood that with Tasmota Thermostat option ME81H features now two thermostats. The original one and the Tasmota one as we can see above. Parameters above the line are related to own ME81H thermostat, parameters below are related to Tasmota thermostat. I also understood that are not communicating between each other. So I'm wondering whether this ME81H unit is capable to be used with Tasmota thermostat software.
It is not hard to route Temperature and Temperature set (ME81H) to Set point temperature and Current temperature (Tasmota thermostat) thanks to some rules. However there are two big problems so far I couldn't figure out :
- Access the internal relay with some TuyaSend command (I could live without it)
- Detect whenever Tasmota Thermostat output changes and send an MQTT stat/topic/value_of_output or set the internal relay accordingly
Here below an excerpt of Tasmota Thermostat log
```
08:40:06.158 THE: Thermostat[0].status.counter_seconds: 0
08:40:06.160 THE: Thermostat[0].status.thermostat_mode: 1
08:40:06.162 THE: Thermostat[0].diag.state_emergency: 0
08:40:06.164 THE: Thermostat[0].diag.output_inconsist_ctr: 0
08:40:06.165 THE: Thermostat[0].status.controller_mode: 0
08:40:06.167 THE: Thermostat[0].status.command_output: 0
08:40:06.169 THE: Thermostat[0].status.status_output: 0
08:40:06.171 THE: Thermostat[0].status.status_input: 1
08:40:06.172 THE: Thermostat[0].status.phase_hybrid_ctr: 1
08:40:06.174 THE: Thermostat[0].status.sensor_alive: 1
08:40:06.176 THE: Thermostat[0].status.status_cycle_active: 0
08:40:06.178 THE: Thermostat[0].temp_pi_error: -6.00 degrees
08:40:06.180 THE: Thermostat[0].temp_pi_accum_error: 0.00 degrees
08:40:06.182 THE: Thermostat[0].time_proportional_pi: 0 seconds
08:40:06.184 THE: Thermostat[0].time_integral_pi: 0 seconds
08:40:06.185 THE: Thermostat[0].time_total_pi: 0 seconds
08:40:06.187 THE: Thermostat[0].temp_measured_gradient: 1728.000 degrees/hour
08:40:06.189 THE: Thermostat[0].time_rampup_deadtime: 0 seconds
08:40:06.191 THE: Thermostat[0].temp_rampup_meas_gradient: 0.000 degrees/hour
08:40:06.192 THE: Thermostat[0].time_ctr_changepoint: 60
08:40:06.194 THE: Thermostat[0].temp_rampup_output_off: 0
08:40:06.196 THE: Thermostat[0].time_ctr_checkpoint: 1860
08:40:06.197 THE: uptime: 420
08:40:06.199 THE: power: 0
```
I noticed that power variable is set to 1 when heating is needed and to 0 when not. Question how to map this to some MQTT send? I indeed noticed the command OutputRelaySet but it is of no use as of my current understanding.

**Comment by g6094199** [comment]
hi @Kolia56 

can you give us the complete me81h template (as of now) cause blackadders doenst seem to be as much as complete as yours.
is it water, boiler or electrical heating version?

thx!

**Reply by barbudor** [reply]
Anyone is welcomed to update the template repository
This lives only by the good will of the community

![image](https://github.com/arendst/Tasmota/assets/2996042/29987f37-9b8e-48db-970d-00818d757a67)

**Reply by Kolia56** [reply]
Unit is Wifi electric heating
![image](https://github.com/arendst/Tasmota/assets/20872098/a6c19459-331c-4c19-9350-514293d6b904)

My template is the very same one as the one in the repository
```
{"NAME":"Thermostat","GPIO":[1,1,1,1,1,1,0,0,1,1,1,1,1,0],"FLAG":0,"BASE":54}
```
There are two power commands available:

- power1 that switches off/on of the unit which is broadcasted as stat/topic/POWER1
- power2 that sets Manual/Auto mode that is broadcasted as stat/topic/POWER2 

But there is no power command to set (which is understandable) or read the state of the internal relay that drives the heater. I managed to get the reading of this relay by doing some reverse engineering. It turns out that using a rule, I get the relay state reported with MQTT
```
{"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":181,"Free":330,"Rules":"on TuyaReceived#Data=55AA03070005240400010037 DO publish stat/tasmota_0E7070/RELAY 1 endon on TuyaReceived#Data=55AA03070005240400010138 DO publish stat/tasmota_0E7070/RELAY 0 endon"}}
```
However (AFAIK) PI feature is of no use, because PI does not drive the relay so far. It would be great if it could, this would solve the issue that is described herebelow.
So now the problem with PI feature. When PI thermostat sets power to 0 or to 1 (see below), Tasmota broadcasts its value to POWER1, making thinking mistakenly that unit is powered On or Off. Instead it should broadcast the value to another variable, like POWER3 for instance.
So as off my knowledge, PI feature cannot be used on the unit. I hope you'll bring some light to this conclusion
```
10:58:19.226 SRC: Thermostat
10:58:19.240 TYA: Send "55aa0006000501010001010e"
10:58:19.265 MQT: stat/tasmota_0E7070/RESULT = {"POWER1":"ON"}
10:58:19.268 MQT: stat/tasmota_0E7070/POWER1 = ON
10:58:19.271 THE: Thermostat[0].status.counter_seconds: 0
10:58:19.272 THE: Thermostat[0].status.thermostat_mode: 1
10:58:19.274 THE: Thermostat[0].diag.state_emergency: 0
10:58:19.276 THE: Thermostat[0].diag.output_inconsist_ctr: 0
10:58:19.278 THE: Thermostat[0].status.controller_mode: 0
10:58:19.279 THE: Thermostat[0].status.command_output: 1
10:58:19.281 THE: Thermostat[0].status.status_output: 1
10:58:19.283 THE: Thermostat[0].status.status_input: 1
10:58:19.285 THE: Thermostat[0].status.phase_hybrid_ctr: 1
10:58:19.286 THE: Thermostat[0].status.sensor_alive: 0
10:58:19.288 THE: Thermostat[0].status.status_cycle_active: 1
10:58:19.290 THE: Thermostat[0].temp_pi_error: 18.00 degrees
10:58:19.292 THE: Thermostat[0].temp_pi_accum_error: 0.00 degrees
10:58:19.294 THE: Thermostat[0].time_proportional_pi: 1800 seconds
10:58:19.296 THE: Thermostat[0].time_integral_pi: 0 seconds
10:58:19.298 THE: Thermostat[0].time_total_pi: 1200 seconds
10:58:19.299 THE: Thermostat[0].temp_measured_gradient: 0.000 degrees/hour
10:58:19.301 THE: Thermostat[0].time_rampup_deadtime: 0 seconds
10:58:19.303 THE: Thermostat[0].temp_rampup_meas_gradient: 0.000 degrees/hour
10:58:19.305 THE: Thermostat[0].time_ctr_changepoint: 60780
10:58:19.307 THE: Thermostat[0].temp_rampup_output_off: 0
10:58:19.308 THE: Thermostat[0].time_ctr_checkpoint: 61380
10:58:19.310 THE: uptime: 60720
10:58:19.312 THE: power: 1
```
One other issue, but I thing I may tackle it using some rules, Set Point Temperature and and current temperature of PI feature are not linked to the related values of the physical thermostat, it would be great to get it native to tasmota. 
I hope that what is exposed is clear, don't hesitate to ask for more information.

**Comment by nikvostro** [comment]
Dears @Kolia56 and @g6094199,
Could you please share with me some experience with ME81H...
1. Which module controls display? What happens with the display once you replace factory MCU with EPS one?
2. @Kolia56 could you please share the pinout you've observed and the Tuya MCU commands you've investigated as well.
3. I have 6.8 kOhm (at 25C) sensor installed which I can't replace unfortunatelly. I suppose it's not an option just to add additional resistor because of different temp-resistance characterisctics comparing to the 10 kOhm one which comes with the ME81H... Could you please give you advices on that?
4. How the externally connected thermoresistor is polled and reported to ESP?

**Comment by Kolia56** [comment]
> 1. Which module controls display? What happens with the display once you replace factory MCU with EPS one?

The display works the same way as before

> 2\. @Kolia56 could you please share the pinout you've observed and the Tuya MCU commands you've investigated as well.

What do you mean by pinout?

> 4\. How the externally connected thermoresistor is polled and reported to ESP?

No clue whether it is pooled or not.

**Reply by nikvostro** [reply]
Thank you!

I've additionally walked through other topics in the internet related to ME81H. Now it's clear for me: ESP is connected to the Tuya MCU and only using GPIOs which are pre-defined in Module 54 ("BASE":54). MissingTyua Rx/Tx (and pins for other components too) definitions in the Template confused me. So Tuya MCU is driving all the components (display, buttons and temp sensor) and acting as a "gateway" for ESP. Sorry for bothering with such a basic quesitons!

Nikolay

**Reply by Kolia56** [reply]
You did not bothered me, and your statement is indeed right. It took me some time to realize that the internal thermostat and the Tasmota Hybrid thermostat are two separate features, the second one is by no mean linked to the hardware of the unit, either display or relays.
As you may have seen I have troubles to use the tasmota thermostat feature (see poste above). No issue with the generic thermostat though. What are you using on your side?
If this may help you and if you are using Home Assistant, here is my MQTT configuration:
```
climate:
  - name: "Thermostat Salle de bains"
    modes:
      - "off"
      - "heat"
    unique_id: "device_pi_id13"
    mode_state_topic: "stat/tasmota_0E7070/POWER1"
    mode_state_template: "{{ 'heat' if value == 'ON' else 'off' }}"
    mode_command_topic: "cmnd/tasmota_0E7070/Power1"
    mode_command_template: >
      {{'on' if value == 'heat' else 'off'}}      
    current_temperature_topic: "tele/tasmota_0E7070/SENSOR"
    current_temperature_template: >
      {% if ( value_json['TuyaSNS'] | regex_search('Temperature') ) %}
        {{ value_json['TuyaSNS'].Temperature}}
      {% endif %}
    temperature_state_topic: "tele/tasmota_0E7070/SENSOR"
    temperature_state_template: >
      {% if ( value_json['TuyaSNS'] | regex_search('TempSet') ) %}
        {{ value_json['TuyaSNS'].TempSet | multiply(10)}}
      {% endif %}
    temperature_command_topic: "cmnd/tasmota_0E7070/TuyaSend2"
    temperature_command_template: "16,{{value | multiply(1)}}"
    action_topic: "stat/tasmota_0E7070/RELAY"
    action_template: >
      {{'heating' if value == '1' else 'idle'}}
    min_temp: 5
    max_temp: 25
    precision: 0.5
    temp_step: 1
    initial: 15
    availability_topic: "tele/tasmota_0E7070/LWT"
    payload_available: "Online"
    payload_not_available: "Offline"
    retain: false
```

**Reply by nikvostro** [reply]
Thanks again!

It could be my first thermostat either BUT just a couple hours ago I realized that the heating module is probably broken (I've newer used heater in my apartments before) - it's resistance just a 26 Ohm and the power wiring and connectors are getting hot much faster than the floor... Heating surface is just about 4-5m^2 hence the resistance of 26 Ohms looks too small.
Now I have to abandon as I'm not planning to dig the floor for rapairing :(

**Reply by joba-1** [reply]
P = U^2/R -> with 230V that is ~ 2kW

So 26 Ohm seems reasonable (but I have no personal experience with those). Still, connectors and wiring getting hot is definitely not good

**Reply by Kolia56** [reply]
as a matter of fact I would never user the thermostat internal relay to switch such a high load but I would use it to switch another high power relay. Doing so, the unit is protected

**Reply by nikvostro** [reply]
Hmmm... I've decided to test again. It's a twisted 1.5mm2 wire used for the heater and 2.5mm2 single wire coming from electrical panel. I did a cleanup and soldered twisted wires for better connection. Connected tightly via screw terminals. Heating is minimal now.
And................ after some minutes I've obtained as huge heater surface but no 4-5m2 I tought before... so 26 Ohm looks much more reasonable now :)

Now I [re-ordered](https://aliexpress.ru/item/4000398480147.html) and will start playing soon. There is also need to add a resistor to the sensor because I have 6.8 kOhm and the subject controller works with 10k. Sensor is sitting too deep and can't be replaced either :(

Will get back too you soon. Or next year maybe :)

![IMG_0051](https://github.com/arendst/Tasmota/assets/83143316/1c35ee64-7503-415f-a726-5184c05b2ed2) ![IMG_0052](https://github.com/arendst/Tasmota/assets/83143316/62a8cf51-f328-47bd-bf8d-68f036ec649e)

**Comment by nikvostro** [comment]
Hi all! Now I have ME81AH on hands!
Got sniffed some data sent from Wi-Fi MCU (Tx pin) to the Tuya MCU. Operations are associated in the moment of activating/changing/pressing some buttons and switches in the Smart Tuya android app.
Looks quite simple to decode. Feel free to attend for decoding! :)

Merry Christmas and a Happy New year!

[Tuya.txt](https://github.com/arendst/Tasmota/files/13834282/Tuya.txt)
